133 integrate movie details api and implement data repositories

hagarabobakr opened the pull request:
Adds Movie repository implementation
DTOs and integrates /movie/{id} endpoint
Introduces MovieRepository (remote)
Type unit tests → Hight overall coverage

hagarabobakr committed:
feat(movies-repository): add dependencies for movies repository

Adds new dependencies and plugins to the `movies` repository module.

The following plugins have been added:
- `ksp` for annotation processing
- `kotlin.serialization` for JSON serialization

The following dependencies have been added:
- `domain.vod` and `envConfig` project modules
- `ktor.serialization.kotlinx.json` for Ktor JSON serialization
- `kotlinx.coroutines.android` for coroutines support
- `junit.jupiter`, `junit.jupiter.api`, `mockk`, `kotlinx.coroutines.test`, `truth`, and `kotlin("test")` for testing

Additionally, JUnit Platform is now used for running tests.

hagarabobakr committed:
add data transfer object for movie details

hagarabobakr committed:
add movie review data models

hagarabobakr committed:
add data transfer object for actor

hagarabobakr committed:
Create MovieDetailsRemoteDataSource interface

This interface defines methods for fetching movie details, images, cast, similar movies, reviews, and movies by category from a remote data source.

hagarabobakr committed:
Add implementation for movie details remote data source

This commit introduces `MovieDetailsRemoteDataSourceImpl`, which implements the `MovieDetailsRemoteDataSource` interface.

The implementation includes methods to:
- Fetch movie details by ID
- Fetch movie images by ID
- Fetch movie cast by ID
- Fetch similar movies by movie ID
- Fetch movie reviews by movie ID
- Fetch movies by category

All methods utilize a `HttpClient` for network requests and incorporate language and API key parameters.

hagarabobakr committed:
add initial mappers for data layer

hagarabobakr committed:
use integer for movie category ID in MovieDetailsRemoteDataSource

hagarabobakr committed:
Setup build configuration and dependencies for movies remote data source

This commit introduces the following changes to the `build.gradle.kts` file for the `data/remote_data_source/movies` module:
- Adds the Kotlin Serialization plugin.
- Configures loading of API keys from `keys.properties`.
- Adds `TMDB_API_KEY` to `buildConfigField`.
- Enables `buildConfig` in `buildFeatures`.
- Includes dependencies for:
    - `envConfig` project module
    - `data.repositories.movies` project module
    - Ktor client libraries
    - Koin for dependency injection
    - Coroutines
    - Firebase Crashlytics build tools
- Adds test dependencies:
    - MockK
    - Kotlinx Coroutines Test
    - Truth
    - Ktor Client Mock
    - Kotlin Test
- Configures tests to use JUnit Platform.

hagarabobakr committed:
remove .gitkeep file

hagarabobakr committed:
Merge branch 'feature/creat-movie-details-usecases' into 133-integrate-movie-details-api-and-implement-data-repositories

hagarabobakr committed:
add biography, character and gender fields to ActorDto

hagarabobakr committed:
implement ActorMapper to map ActorDto to Actor and vice versa

hagarabobakr committed:
add kotlinx.datetime dependency

hagarabobakr committed:
remove unnecessary .gitkeep files from movies and search modules

hagarabobakr committed:
add `toGenre()` function for converting integer IDs to `Genre` enum

hagarabobakr committed:
add gender enum in ActorDto to ensure type safety

hagarabobakr committed:
add a mapper function to convert `ReviewDto` objects to `Review` domain entities.

hagarabobakr committed:
add mapper from movie details dto to movie entity

hagarabobakr committed:
implement MovieRepository in data layer
the implementation includes methods to:
- Get movie details
- Get movie images
- Get movie cast
- Get similar movies
- Get movie reviews
- Get movies by category

hagarabobakr committed:
add networkModule to di package

hagarabobakr committed:
add remoteDataSource to di package

hagarabobakr committed:
update imports

hagarabobakr committed:
Merge remote-tracking branch 'origin/development' into 133-integrate-movie-details-api-and-implement-data-repositories

hagarabobakr committed:
enhance: refine DTOs for movie details and related data

- Add `CastDto`, `MovieImagesDto`, and `SimilarMoviesDto` to handle new API responses.
- Update `ActorDto`, `MovieDetailsDto`, and `ReviewDto` with additional fields and improved structure.
- Move DTOs to the `dataSource.remote.dto` package.

hagarabobakr committed:
add new modules to app dependencies

The modules for movies repository and movies remote data source are added as dependencies to the `app/build.gradle.kts` file.

hagarabobakr committed:
update mapper classes

this commit updates the `ActorMapper`, `MovieMapper`, and `ReviewMapper` classes.

hagarabobakr committed:
refactor MovieDetailsRemoteDataSource and implementation

hagarabobakr committed:
update MovieRepositoryImpl with error handling and data mapping

hagarabobakr committed:
inject MovieRepositoryImpl

This commit injects `MovieRepositoryImpl` into the application using Koin.
It also updates related DI modules (`appModule`, `domainModule`, `repositoryModule`, and `remoteMovieDetailsDataSource`) to include the necessary dependencies for movie details functionality.

hagarabobakr committed:
remove networkModule

hagarabobakr committed:
add MovieRepositoryImpl tests

hagarabobakr committed:
add unit tests for actor mapper

hagarabobakr committed:
add tests for GenreMapper

hagarabobakr committed:
add tests for ReviewMapper

hagarabobakr committed:
add test cases for MovieMapper

hagarabobakr committed:
add MovieDetailsRemoteDataSourceImplTest

hagarabobakr committed:
remove unused imports

hagarabobakr committed:
Merge remote-tracking branch 'refs/remotes/origin/development' into 133-integrate-movie-details-api-and-implement-data-repositories

hagarabobakr committed:
remove firebase-crashlytics-buildtools dependency

Radwa85 commented on data/remote_data_source/movies/src/test/java/com/sanaa/movies/MovieDetailsRemoteDataSourceImplTest.kt:
```diff
@@ -0,0 +1,118 @@
+        val dto = dataSource.fetchSimilarMoviesByMovieId(1)
+        assertEquals(1, dto.page)
+        assertEquals(1, dto.results.size)
+        assertEquals(2, dto.results.first().id)
```
every test should have one assert equals

Radwa85 commented on data/remote_data_source/movies/src/test/java/com/sanaa/movies/MovieDetailsRemoteDataSourceImplTest.kt:
```diff
@@ -0,0 +1,118 @@
+        headersOf(HttpHeaders.ContentType, ContentType.Application.Json.toString())
+    )
+
+    @Test fun `fetchMovieDetails returns correct DTO`() = runTest {
```
The @Test annotation should not be placed on the same line as the function declaration for better readability and consistency

Radwa85 commented on data/remote_data_source/movies/src/test/java/com/sanaa/movies/MovieDetailsRemoteDataSourceImplTest.kt:
```diff
@@ -0,0 +1,118 @@
+        assertEquals(100, dto.duration)
+    }
+
+    @Test fun `fetchImages returns correct MovieImagesDto`() = runTest {
```
enhance test readability using should ...when

Radwa85 commented on data/repositories/movies/src/main/java/com/sanaa/movies/repository/MovieRepositoryImpl.kt:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    override suspend fun getMoviesByCategory(category: Genre): List<Movie> {
+        TODO()
```
i think u forget it

Ahmed7osny1 replied:
@hagarabobakr what about this ?

hagarabobakr replied:
I forgot it and I'm currently writing it down and I will push it.

Radwa85 commented on data/repositories/movies/src/test/java/com/sanaa/movies/repository/MovieRepositoryImplTest.kt:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    @Test
+    fun `getMovieCast returns cast list`() = runTest {
```
also here use should... when

Radwa85 commented on data/remote_data_source/movies/src/main/java/com/sanaa/movies/MovieDetailsRemoteDataSourceImpl.kt:
```diff
@@ -0,0 +1,57 @@
+    ): T = client.get("$baseUrl/$path") {
+        if (withLang) parameter("language", languageProvider.getCurrentLanguage())
+        parameter("api_key", BuildConfig.TMDB_API_KEY)
+    }.body()
```
make private function at bottom of class (one of review tips)

AhmadKharfan committed:
refactor ktor body deserialization for jacoco coverage

hagarabobakr committed:
improve unit test coverage for genre mappers

hagarabobakr committed:
add invalid id returns null genre test

hagarabobakr committed:
improve tests for movie mapper

The commit improves the test coverage for the `MovieMapper` by:

- Separating tests for `MovieDetailsDto` and `SimilarMoviesDtoResults` mapping.
- Adding specific tests for each mapped field (id, genres, imdbRating, posterImageUrl, title).
- Introducing helper functions (`createSampleMovieDetailsDto` and `createSampleSimilarMovieDto`) to provide sample DTOs for testing, making tests cleaner and more focused.

hagarabobakr committed:
add unit tests for movie mapper with default values

hagarabobakr committed:
split unit tests in MovieDetailsRemoteDataSourceImplTest

hagarabobakr committed:
Merge branch 'development' into 133-integrate-movie-details-api-and-implement-data-repositories

Ahmed7osny1 commented on data/repositories/movies/src/main/java/com/sanaa/movies/mapper/ActorMapper.kt:
```diff
@@ -0,0 +1,51 @@
+
+fun apiGenderMapping(id: Int?): Gender {
+    return when (id) {
+        0 -> Gender.MALE
+        1 -> Gender.FEMALE
+        else -> Gender.MALE
```
You need to avoid using magic numbers

Ahmed7osny1 commented on data/repositories/movies/src/main/java/com/sanaa/movies/mapper/ActorMapper.kt:
```diff
@@ -0,0 +1,51 @@
+fun CastDto.Cast.toDomain(): Actor {
+
+    return Actor(
+        id = id ?: 0,
```
How id value be null?

Ahmed7osny1 commented on data/repositories/movies/src/test/java/com/sanaa/movies/mapper/ReviewMapperKtTest.kt:
```diff
@@ -0,0 +1,32 @@
+        assertThat(result.userHandle).isEqualTo("johnd")
+        assertThat(result.avatarUrl).contains("/avatar.jpg")
+        assertThat(result.rating).isEqualTo(8.0f)
+        assertThat(result.content).isEqualTo("Amazing movie!")
```
you can create model and compare result with this model

Ahmed7osny1 commented on data/repositories/movies/src/test/java/com/sanaa/movies/repository/MovieRepositoryImplTest.kt:
```diff
@@ -0,0 +1,138 @@
+            totalResults = 2
+        )
+
+        private val sampleReviewDto = ReviewDto(
+            id = 1,
+            page = 1,
+            totalPages = 1,
+            totalResults = 2,
+            results = arrayListOf(
+                ReviewDto.Results(author = "Critic A", content = "Nice"),
+                ReviewDto.Results(author = "Critic B", content = "Meh")
+            )
```
@hagarabobakr like here

raouf-kamel commented on data/repositories/movies/src/main/java/com/sanaa/movies/repository/MovieRepositoryImpl.kt:
```diff
@@ -0,0 +1,75 @@
+    private val remote: MovieDetailsRemoteDataSource
+) : MovieRepository {
+    override suspend fun getMovieDetails(id: Int): Movie {
+        try {
```
I think it is better to create a try function for everyone.

Ahmed7osny1 commented on data/repositories/movies/src/test/java/com/sanaa/movies/mapper/ActorMapperKtTest.kt:
```diff
@@ -0,0 +1,49 @@
+
+        val result = dto.toDomain()
+
+        assertThat(result.id).isEqualTo(1)
+        assertThat(result.name).isEqualTo("Robert Downey Jr.")
+        assertThat(result.imageUrl).isEqualTo("/rdj.jpg")
+        assertThat(result.gender).isEqualTo(Actor.Gender.MALE)
+        assertThat(result.character).isEqualTo("Iron Man")
+        assertThat(result.biography).isEqualTo("Bio")
```
the same here

raouf-kamel commented on data/repositories/movies/src/test/java/com/sanaa/movies/mapper/ReviewMapperKtTest.kt:
```diff
@@ -0,0 +1,32 @@
+
+        val result = dto.toDomain()
+
+        assertThat(result.id).isEqualTo(10)
```
I think it is better not to have more than one verification in the same function.

Ahmed7osny1 commented on data/repositories/movies/src/main/java/com/sanaa/movies/repository/MovieRepositoryImpl.kt:
```diff
@@ -0,0 +1,75 @@
+    private val remote: MovieDetailsRemoteDataSource
+) : MovieRepository {
+    override suspend fun getMovieDetails(id: Int): Movie {
+        try {
+            return remote.fetchMovieDetails(id).toDomain()
+        } catch (_: UnknownHostException) {
+            throw NoNetworkException()
+        } catch (e: Exception) {
+            throw RetrievingDataFailureException("Failed to fetch movie details: ${e.message}")
+        }
```
try catch frequently, you can use a function to avoid this

raouf-kamel reviewed (dismissed):
💪 **Keep it up, hagar**

Ahmed7osny1 commented on data/repositories/movies/src/main/java/com/sanaa/movies/mapper/ReviewMapper.kt:
```diff
@@ -0,0 +1,26 @@
+
+fun ReviewDto.Results.toDomain(): Review {
+    return Review(
+        id = id?.toInt() ?: 0,
```
i don't understand how id return with null value ?

Ahmed7osny1 commented on data/repositories/movies/src/main/java/com/sanaa/movies/dataSource/remote/MovieDetailsRemoteDataSource.kt:
```diff
@@ -0,0 +1,16 @@
+
+interface MovieDetailsRemoteDataSource {
+    suspend fun fetchMovieDetails(id: Int): MovieDetailsDto
+    suspend fun fetchImages(id: Int): MovieImagesDto
```
fetch image url ?

Radwa85 commented on app/src/main/java/com/sanaa/novix/di/repositoryModule.kt:
```diff
@@ -18,5 +21,7 @@ val repositoryModule = module {
     single<LocalSearchHistoryDataSource> { LocalSearchHistoryDataSourceImpl(get(), get()) }
 
     single<ActorRepository> { ActorRepositoryImpl(get()) }
+    single<MovieRepositoryImpl> { MovieRepositoryImpl(get()) }
```
You should register interfaces not implementations in modules

Radwa85 replied:
u should register interfaces,not implementations in module

hagarabobakr committed:
refactor after review

hagarabobakr committed:
Merge branch 'refs/heads/development' into 133-integrate-movie-details-api-and-implement-data-repositories

hagarabobakr committed:
refactor after review
